### PR TITLE
Show finalizing devworkspaces until it's actually gone

### PR DESCRIPTION
**Pull request by olexii4**
Signed-off-by: Oleksii Orel <oorel@redhat.com>

### What does this PR do?
Show finalizing devworkspaces until it's actually gone

### What issues does this PR fix or reference?
fixed https://github.com/eclipse/che/issues/19324

### CODE REVIEW

File: src/services/workspace-client/devWorkspaceClient.ts [hunk 1]
```diff
@@ -67,7 +67,7 @@ export class DevWorkspaceClient extends WorkspaceClient {
     const workspaces = await this.dwApi.listInNamespace(defaultNamespace);
     const availableWorkspaces: IDevWorkspace[] = [];
     for (const workspace of workspaces) {
-      if (!isDeleting(workspace) && !isWebTerminal(workspace)) {
```
Comment (sleshchenko) [inline src/services/workspace-client/devWorkspaceClient.ts:70]: The import needs to be cleaned up, after you remove usage. Actually, isDeleting func is totally unused now.

File: src/services/helpers/types.ts [hunk 2]
```diff
@@ -28,16 +28,17 @@ export interface FactoryResolver {
...
+  STOPPING = 'STOPPING',
+  STOPPED = 'STOPPED',
+  STARTING = 'STARTING',
+  PAUSED = 'PAUSED',
+  ERROR = 'ERROR',
 }
 
 export enum DevWorkspaceStatus {
   FAILED = 'Failed',
+  STARTING = 'Starting',
```
Comment (sleshchenko) [inline src/services/helpers/types.ts:41]: It seems to miss terminating state, and as result I assume Terminating workspaces are not displayed, to check that I've added some logs:
Screenshot_20210427_121625

Also, note that after I removed Starting devworkspace, I got the warning in browser logs:
Screenshot_20210427_121704

**Review by sleshchenko — COMMENTED** [review]
it does not seem to work yet

File: src/services/workspace-client/devWorkspaceClient.ts (hunk 1 above)
Comment (olexii4) [reply src/services/workspace-client/devWorkspaceClient.ts:70]: I have fixed it

Comment (olexii4) [reply src/services/workspace-client/devWorkspaceClient.ts:70]: https://github.com/eclipse-che/che-dashboard/pull/225/files#diff-bd3e628ea1cf6067bd9ad9b7dd8509178ad6b95d2593f3f5affd3e673f01033cL19

File: src/services/helpers/types.ts (hunk 2 above)
Comment (olexii4) [reply src/services/helpers/types.ts:41]: I have fixed it

File: src/services/helpers/types.ts [hunk 3]
```diff
@@ -28,16 +28,18 @@ export interface FactoryResolver {
...
-  STOPPING,
-  STOPPED,
-  STARTING,
-  PAUSED,
-  ERROR,
+  RUNNING = 'RUNNING',
+  STOPPING = 'STOPPING',
+  STOPPED = 'STOPPED',
+  STARTING = 'STARTING',
+  PAUSED = 'PAUSED',
```
Comment (sleshchenko) [inline src/services/helpers/types.ts:35]: Do we really have paused for Che Workspaces, I don't think so.

File: src/store/Workspaces/devWorkspaces/index.ts [hunk 4]
```diff
@@ -71,6 +73,11 @@ interface DeleteWorkspaceAction extends Action {
   workspaceId: string;
 }
 
+interface TerminateWorkspaceAction extends Action {
+  type: 'DEV_TERMINATE_WORKSPACE';
```
Comment (sleshchenko) [inline src/store/Workspaces/devWorkspaces/index.ts:77]: Is this DEV means dev event or DevWorkspaces. If the second I would rename it like the following
```suggestion
  type: 'TERMINATE_DEVWORKSPACE';
```

File: src/services/helpers/types.ts (hunk 3 above)
Comment (olexii4) [reply src/services/helpers/types.ts]: I have fixed it

File: src/store/Workspaces/devWorkspaces/index.ts (hunk 4 above)
Comment (olexii4) [reply src/store/Workspaces/devWorkspaces/index.ts]: I have fixed it

**Review by sleshchenko — COMMENTED** [review]
The stuff with an ability to restart failed DevWorkspace seems to be critical. Not approving yet

File: src/pages/WorkspacesList/index.tsx [hunk 5]
```diff
@@ -151,50 +144,37 @@ export default class WorkspacesList extends React.PureComponent<Props, State> {
     return [
       {
         title: 'Open in Verbose Mode',
-        isDisabled: false === this.isEnabledAction(WorkspaceAction.START_DEBUG_AND_OPEN_LOGS, WorkspaceStatus[workspace.status]),
+        isDisabled: !this.isEnabledAction(WorkspaceAction.START_DEBUG_AND_OPEN_LOGS, workspace),
         onClick: (event, rowId, rowData) => this.handleAction(WorkspaceAction.START_DEBUG_AND_OPEN_LOGS, rowData),
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply src/pages/WorkspacesList/index.tsx:148]: Checked, I can reproduce it on main branch. I'll create a dedicated issue.

**Review by JPinkney — COMMENTED** [review]
The code generally looks OK, I'm going to wait to approve until Serhii's issues are fixed

File: src/store/Workspaces/devWorkspaces/index.ts (hunk 4 above)
Comment (akurinnoy) [reply src/store/Workspaces/devWorkspaces/index.ts]: @olexii4 I'd ask you to rename all remaining actions to keep their consistency.

Comment (olexii4) [reply src/store/Workspaces/devWorkspaces/index.ts]: I have fixed it

File: src/pages/WorkspacesList/Rows.tsx [hunk 6]
```diff
@@ -134,7 +135,7 @@ export function buildRow(
 
   /* Open IDE link */
   let open: React.ReactElement | string;
-  if (isDeleted) {
+  if (deleting || workspace.status === DevWorkspaceStatus.TERMINATING) {
     open = 'deleting...';
```
[the remark this answers is not in the document]
Comment (olexii4) [reply src/pages/WorkspacesList/Rows.tsx:139]: I have fixed it

**Review by sleshchenko — APPROVED** [review]
I haven't retested but I believe now it should be possible to restart failed devworkspace

File: src/pages/GetStarted/CustomWorkspaceTab/index.tsx [hunk 7]
```diff
@@ -145,14 +146,19 @@ export class CustomWorkspaceTab extends React.PureComponent<Props, State> {
...
   }
 
-  private handleCreate(): Promise<void> {
-    return this.props.onDevfile(this.state.devfile, this.state.namespace?.name);
+  private handleCreate(): void {
+    this.setState({ hasCreated: true });
+    const promise = this.props.onDevfile(this.state.devfile, this.state.namespace?.name);
+    if (promise) {
+      promise.catch(() => this.setState({ hasCreated: false }));
+    }
```
[the remark this answers is not in the document]
Comment (olexii4) [reply src/pages/GetStarted/CustomWorkspaceTab/index.tsx]: I have fixed it.

File: src/containers/IdeLoader.tsx [hunk 8]
```diff
@@ -134,6 +134,37 @@ class IdeLoaderContainer extends React.PureComponent<Props, State> {
     }
   }
 
+  private async handleEventMessage(event: MessageEvent): Promise<void> {
```
[the remark this answers is not in the document]
Comment (olexii4) [reply src/containers/IdeLoader.tsx]: I have fixed it.

**Review by akurinnoy — APPROVED** [review]

**Pull request merged**